### ISSUE TITLE
[V3 Utils/Mod] Mod helpers allow users to be passed, raising AttributeError

### ISSUE THREAD

**Issue by Tobotimus** [Cog-Creators/Red-DiscordBot#1202]
### Type:

- Bug

### Brief description of the problem
Following #1185, `is_admin_or_superior` and `is_mod_or_superior` allow `discord.User` to be passed in, but when this occurs, `AttributeError` is raised.

**UPDATE:** these functions really shouldn't be used for `discord.User` at all, since it doesn't make any logical sense to check for permissions on a `User` object without any guild. So to resolve this issue we should really be type-checking the arguments to the function before using it, i.e. in mod.py and filter.py.

### Expected behavior
Helpers should raise `TypeError` (as it was prior to #1185)

### Actual behavior
```py
AttributeError: 'User' object has no attribute 'guild'
  File "discord/client.py", line 223, in _run_event
    yield from coro(*args, **kwargs)
  File "filter/filter.py", line 266, in on_message_edit
    mod_or_superior = await is_mod_or_superior(self.bot, obj=author)
  File "redbot/core/utils/mod.py", line 134, in is_mod_or_superior
    server = obj.guild
```

### Steps to reproduce

1. Pass a `discord.User` into `is_admin_or_superior` or `is_mod_or_superior`

**Comment by Tobotimus** [comment]
[Corresponding sentry issue](https://sentry.io/will-tekulve/bot-development/issues/420900834/)

**Comment by tekulvw** [comment]
Fixed in #1327
